src_fun/main.cpp

```diff
@@ -3,46 +3,110 @@
 #include "../src_common/hash.h"
 #include "../src_common/integer.h"
 #include "api.h"
+#include "testing.h"
+
 
 EnterApplicationNamespace
 
 
+static const String TEST_DIRECTORY = "test"_s;
+
+
 extern "C" int main(int argc, char** argv)
 {
     /*add_log_handler([](String severity, String subsystem, String msg)
     {
         printf("%.*s\n", StringArgs(msg));
     });*/
 
-    Dynamic_Array<String> non_flag_args = {};
-    for (umm i = 1; i < argc; i++)
-        if (argv[i][0] != '-')
-            *reserve_item(&non_flag_args) = wrap_string(argv[i]);
+    String first_flag_arg = {};
+    if (argc > 1 && argv[1][0] == '-')
+        first_flag_arg = consume(wrap_string(argv[1]), 1);
+
+    if (first_flag_arg == "test_runner"_s)
+        return test_runner_entry();
 
-    if (non_flag_args.count != 1 && non_flag_args.count != 2)
+    if (first_flag_arg == "test"_s)
+    {
+        add_log_handler([](String severity, String subsystem, String msg)
+        {
+            printf("%.*s\n", StringArgs(msg));
+        });
+
+        Testing_Context context = {};
+        Defer(free_test_suite(&context));
+        if (!initialize_test_suite(&context, TEST_DIRECTORY)) return 1;
+        run_tests(&context, argv[0]);
+
+        return 0;
+    }
+
+    if (first_flag_arg == "stress_test"_s)
+    {
+        Testing_Context context = {};
+        Defer(free_test_suite(&context));
+        if (!initialize_test_suite(&context, TEST_DIRECTORY)) return 1;
+
+        umm iteration = 0;
+        while (true)
+        {
+            if (!run_tests(&context, argv[0], /* only log fails */ true))
+            {
+                printf("\n\nSome tests started failing after %lu iterations.\n", iteration);
+                break;
+            }
+
+            iteration++;
+            if (iteration % 1000 == 0)
+                printf("Ran all tests %lu times.\n", iteration);
+        }
+
+        return 0;
+    }
+
+    if (first_flag_arg == "run"_s)
+    {
+        String_Concatenator code_cat = {};
+        add(&code_cat, "run unit { debug\n"_s);
+        for (umm i = 2; i < argc; i++)
+        {
+            add(&code_cat, " "_s);

```

Comment:
Do not add this space before the first argument. You will offset the column of tokens by 1 relative to the start of the argument, which is just bad error reporting.